### ISSUE TITLE
Comments break Fluent (FTL) serialization

### ISSUE THREAD

**Issue by flaki** [translate/translate#4429]
The current implementation seems to completely choke on any [free-standing Fluent comments](https://projectfluent.org/fluent/guide/comments.html) during serialization. Weblate imports the FTL seemingly fine, but exporting in any way (repository commits or file downloads) results in a fatal error:

- `TypeError - 'NoneType' object is not subscriptable`
- Traced to `"/home/weblate/weblate-env/lib/python3.8/site-packages/fluent/syntax/stream.py", line 13, in get return self.string[offset]`

Detailed error log:

<details>

```
Internal Server Error: /download/rustfest-cfp/cfp-form/en/

TypeError at /download/rustfest-cfp/cfp-form/en/
'NoneType' object is not subscriptable

Request Method: GET
Request URL: http://weblate.rustfest.global/download/rustfest-cfp/cfp-form/en/
Django Version: 3.2.7
Python Executable: /home/weblate/weblate-env/bin/uwsgi
Python Version: 3.8.10
Python Path: ['/', '/home/weblate/weblate-env/lib/python3.8/site-packages/git/ext/gitdb', '.', '', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '/home/weblate/weblate-env/lib/python3.8/site-packages', '/home/weblate/weblate-env/lib/python3.8/site-packages/gitdb/ext/smmap']
Server time: Wed, 08 Sep 2021 17:59:26 +0000
Installed Applications:
['weblate.addons',
 'weblate.auth',
 'weblate.checks',
 'weblate.formats',
 'weblate.glossary',
 'weblate.machinery',
 'weblate.trans',
 'weblate.lang',
 'weblate_language_data',
 'weblate.memory',
 'weblate.screenshots',
 'weblate.fonts',
 'weblate.accounts',
 'weblate.configuration',
 'weblate.utils',
 'weblate.vcs',
 'weblate.wladmin',
 'weblate.metrics',
 'weblate',
 'weblate.gitexport',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'django.contrib.admin.apps.SimpleAdminConfig',
 'django.contrib.admindocs',
 'django.contrib.sitemaps',
 'django.contrib.humanize',
 'social_django',
 'crispy_forms',
 'compressor',
 'rest_framework',
 'rest_framework.authtoken',
 'django_filters']
Installed Middleware:
['weblate.middleware.RedirectMiddleware',
 'weblate.middleware.ProxyMiddleware',
 'django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'weblate.accounts.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'social_django.middleware.SocialAuthExceptionMiddleware',
 'weblate.accounts.middleware.RequireLoginMiddleware',
 'weblate.api.middleware.ThrottlingMiddleware',
 'weblate.middleware.SecurityMiddleware',
 'weblate.wladmin.middleware.ManageMiddleware']


Traceback (most recent call last):
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/weblate/trans/views/files.py", line 136, in download_translation
    return download_translation_file(request, obj, **kwargs)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/django/utils/decorators.py", line 130, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/weblate/utils/views.py", line 312, in download_translation_file
    translation.commit_pending("download", None)
  File "/usr/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/weblate/trans/models/translation.py", line 559, in commit_pending
    self.update_units(units, store, author_name, author.id)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/weblate/trans/models/translation.py", line 733, in update_units
    store.save()
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/weblate/formats/ttkit.py", line 304, in save
    self.save_atomic(self.storefile, self.save_content)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/weblate/formats/base.py", line 314, in save_atomic
    callback(temp)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/weblate/formats/ttkit.py", line 300, in save_content
    self.store.serialize(handle)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/translate/storage/fluent.py", line 194, in serialize
    body = [unit.to_entry() for unit in self.units]
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/translate/storage/fluent.py", line 194, in <listcomp>
    body = [unit.to_entry() for unit in self.units]
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/translate/storage/fluent.py", line 151, in to_entry
    value = fp.maybe_get_pattern(FluentParserStream(self.source))
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/fluent/syntax/parser.py", line 343, in maybe_get_pattern
    ps.peek_blank_inline()
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/fluent/syntax/stream.py", line 70, in peek_blank_inline
    while self.current_peek == ' ':
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/fluent/syntax/stream.py", line 34, in current_peek
    return self.char_at(self.index + self.peek_offset)
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/fluent/syntax/stream.py", line 22, in char_at
    if self.get(offset) == '\r' \
  File "/home/weblate/weblate-env/lib/python3.8/site-packages/fluent/syntax/stream.py", line 13, in get
    return self.string[offset]

Exception Type: TypeError at /download/rustfest-cfp/cfp-form/en/
Exception Value: 'NoneType' object is not subscriptable
```

</details>

[Here](https://github.com/RustFestEU/cfp/blob/709e4d5bdd14c2f33b9d1a08401b5bc901ae0fb2/locales/en/cfp.ftl) is the imported original FTL file that should reproduce the issue.

I am not a Python coder but I managed to trace this to handling of various comment nodes (`ast.Resource|Group|Comment`) which are comments used to partition the resource file, as opposed to being descriptions attached to a particular message. I managed to patch `translate/storage/Fluent.ftl` with a fix locally which should help fixing this properly: [The changes](https://gist.github.com/flaki/326736cc8d7b5930e7fed2fcaf6b3896#file-translate_storage_fluent-py-L143-L174) require producing the correct AST nodes for the comment sections so the serialization succeeds.

This is purely a syntactical-fix, without taking into account the semantics of these comments, but it works well for well-formed files at least when importing and exporting the comments.

**Comment by nijel** [comment]
@str4d can you please check this?

**Comment by str4d** [comment]
Thanks for testing this! I've addressed the immediate serialization bug in #4431.
